service/service.go

```diff
@@ -1,272 +1,390 @@
+		if !ok {
+			return errors.New("not an array")
+		}
+		for _, x := range array {
+			if err := p.validateType(x); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	return p.validateType(arg)
 }
 
-// validateConfigurationEnv checks presence of env variables in mesg.yml under env section.
-func (s *Service) validateConfigurationEnv(env map[string]string) error {
-	var nonDefined []string
-	for key := range env {
-		exists := false
-		// check if "key=" exists in configuration
-		for _, env := range s.configuration().Env {
-			if strings.HasPrefix(env, key+"=") {
-				exists = true
-			}
+const (
+	paramStringType  = "String"
+	paramNumberType  = "Number"
+	paramBooleanType = "Boolean"
+	paramOjbectType  = "Object"
+	paramAnyType     = "Any"
+)
+
+// validateType checks if arg comforts its expected type.
+func (p *Parameter) validateType(arg interface{}) error {
+	switch p.Type {
+	case paramStringType:
+		if _, ok := arg.(string); !ok {
+			return errors.New("not a string")
 		}
-		if !exists {
-			nonDefined = append(nonDefined, key)
+	case paramNumberType:
+		if _, ok := arg.(float64); !ok {
+			return errors.New("not a number")
 		}
-	}
-	if len(nonDefined) > 0 {
-		sort.Strings(nonDefined)
-		return ErrNotDefinedEnv{nonDefined}
+	case paramBooleanType:
+		if _, ok := arg.(bool); !ok {
+			return errors.New("not a boolean")
+		}
+	case paramOjbectType:
+		obj, ok := arg.(map[string]interface{})
+		if !ok {
+			return errors.New("not an object")
+		}
+		return validateParametersSchema(p.Object, obj)
+	case paramAnyType:
+		return nil
+	default:
+		return errors.New("not valid type")
 	}
 	return nil
 }
 
-// helper to return the configuration of the service from the dependencies array
-func (s *Service) configuration() *Dependency {
-	for _, dep := range s.Dependencies {
-		if dep.Key == importer.ConfigurationDependencyKey {
-			return dep
+// calculate returns a hash according to the given data.
+func calculate(data []string) string {
+	return strings.Join(data, ".")
+}
+
+// ValidateParametersSchema validates data to see if it matches with parameters schema.
+func validateParametersSchema(parameters map[string]*Parameter, args map[string]interface{}) error {
+	for key, param := range parameters {
+		if err := param.Validate(args[key]); err != nil {
+			return fmt.Errorf("argument %s is %s", key, err)
 		}
 	}
 	return nil
 }
 
-// ErrNotDefinedEnv error returned when optionally given env variables
-// are not defined in the mesg.yml file.
-type ErrNotDefinedEnv struct {
-	env []string
-}
-
-func (e ErrNotDefinedEnv) Error() string {
-	return fmt.Sprintf("environment variable(s) %q not defined in mesg.yml (under configuration.env key)",
-		strings.Join(e.env, ", "))
+// volumeKey creates a key for service's volume based on the sid to make sure that the volume
+// will stay the same for different versions of the service.
+func volumeKey(sid, dependency, volume string) string {
+	h := sha1.New()
+	h.Write([]byte(sid))
```

Comment:
The sid that is now optional is still used here.
I suggest to add a fallback to hash if sid is empty, but hash need to be passed as parameter as well.

```suggestion
	if sid != "" {
		h.Write([]byte(sid))
	} else {
		h.Write([]byte(hash))
	}
```